### PR TITLE
Update Camera and Shadertoy examples

### PR DESCRIPTION
**Pull request by zotho**
- [`get_down_keys`](https://github.com/not-fl3/macroquad/pull/25/files#diff-70b83a8ee1aac5c3dc63f9734657291cR33) for efficent handle down keys

- Updated `camera` example for visualize transitions: target/rotation/zoom/offset in `Camera2D`

- Updated `shadertoy` example for autocompile shader once it edited

- [x] Merge after https://github.com/not-fl3/megaui/pull/12 been merged

### CODE REVIEW

File: examples/camera.rs [hunk 1]
```diff
@@ -1,23 +1,121 @@
...
+
+fn window_conf() -> Conf {
+    Conf {
+        window_title: "Camera".to_string(),
+        fullscreen: true,
+        ..Default::default()
+    }
+}
+
+#[macroquad::main(window_conf)]
```
Comment (not-fl3) [inline examples/camera.rs:30]: Nice, I really like how it looks! :)

File: Cargo.toml [hunk 2]
```diff
@@ -23,3 +23,6 @@ glam = {version = "0.8", features = ["scalar-math"] }
 image = { version = "0.22", default-features = false, features = ["png_codec", "tga"] }
 megaui = { version = "=0.2.11" }
 macroquad_macro = { version = "0.1", path = "macroquad_macro" }
+
+[patch.crates-io]
```
Comment (not-fl3) [inline Cargo.toml:27]: This probably should not be a part of this PR?

File: examples/shadertoy.rs [hunk 3]
```diff
@@ -4,9 +4,17 @@ use macroquad::megaui::widgets::{Label, TreeNode};
...
+        window_title: "Shadertoy".to_string(),
+        fullscreen: true,
+        ..Default::default()
+    }
+}
+
+#[macroquad::main(window_conf)]
 async fn main() {
-    let ferris = load_texture("rust.png").await;
+    let ferris = load_texture("examples/rust.png").await;
```
Comment (not-fl3) [inline examples/shadertoy.rs:17]: IIRC examples/rust.png will not work on wasm - there is no examples folder for textures

File: macroquad_macro/src/lib.rs [hunk 4]
```diff
@@ -34,12 +34,19 @@ pub fn main(attr: TokenStream, item: TokenStream) -> TokenStream {
     }
     modified.extend(source);
 
+    let (method, ident) = match attr.into_iter().next() {
```
Comment (not-fl3) [inline macroquad_macro/src/lib.rs:37]: Could you make separate PR with just macro changes? 
It would be way easier to review!

File: examples/shadertoy.rs (hunk 3 above)
Comment (zotho) [reply examples/shadertoy.rs:17]: In this case have to run executables from examples directory.
Otherwise just run `cargo run --example shadertoy` works correctly

File: macroquad_macro/src/lib.rs (hunk 4 above)
Comment (zotho) [reply macroquad_macro/src/lib.rs:37]: Ok

File: Cargo.toml (hunk 2 above)
Comment (zotho) [reply Cargo.toml]: Removed

File: macroquad_macro/src/lib.rs (hunk 4 above)
Comment (zotho) [reply macroquad_macro/src/lib.rs]: Move macro changes to https://github.com/not-fl3/macroquad/pull/28 PR

File: examples/shadertoy.rs (hunk 3 above)
Comment (zotho) [reply examples/shadertoy.rs:9]: Also there an issue of this problem https://github.com/not-fl3/macroquad/issues/26

File: examples/camera.rs [hunk 5]
```diff
@@ -1,23 +1,113 @@
...
 
-        // Render some primitives in camera space
+    let mut target = (0., 0.);
+    let mut zoom = 1.0;
+    let mut rotation = 0.0;
+    let mut smooth_rotation: f32 = 0.0;
+    let mut offset = (0., 0.);
+
+    'main: loop {
+        for &key in get_down_keys(&mut down_keys).iter() {
```
Comment (not-fl3) [inline examples/camera.rs:33]: I am not quite sold on the idea of `get_down_keys`

As for me, this example would be a bit more readable with just a number of 

```
if is_key_down(Key::Whatever) {
  target.1 -= 0.1;
}
```

File: examples/camera.rs [hunk 6]
```diff
@@ -1,23 +1,113 @@
...
+                KeyCode::Q | KeyCode::Escape => break 'main,
+                _ => ()
+            }
+        }
+
+        match mouse_wheel() {
+            (_x, y) if y != 0.0 => {
+                // Normalise mouse wheel values is browser (chromium: 53, firefox: 3)
+                #[cfg(target_arch = "wasm32")]
+                let y = if y < 0.0 {-1.0} else if y > 0.0 {1.0} else {0.0};
```
Comment (not-fl3) [inline examples/camera.rs:53]: It looks like `rustfmt` would not be happy with this line

File: examples/camera.rs [hunk 7]
```diff
@@ -1,23 +1,113 @@
...
+                KeyCode::W => target.1 -= 0.1,
+                KeyCode::S => target.1 += 0.1,
+                KeyCode::A => target.0 += 0.1,
+                KeyCode::D => target.0 -= 0.1,
+                KeyCode::Left => offset.0 -= 0.1,
+                KeyCode::Right => offset.0 += 0.1,
+                KeyCode::Up => offset.1 += 0.1,
+                KeyCode::Down => offset.1 -= 0.1,
+                #[cfg(not(target_arch = "wasm32"))]
+                KeyCode::Q | KeyCode::Escape => break 'main,
```
Comment (not-fl3) [inline examples/camera.rs:44]: I think we will need some `exit` mechanism in macroquad. 
Until it's done - I would try to avoid 'main on loops to not confuse people too much :)

File: examples/camera.rs [hunk 8]
```diff
@@ -1,23 +1,113 @@
...
+                _ => ()
+            }
+        }
+
+        match mouse_wheel() {
+            (_x, y) if y != 0.0 => {
+                // Normalise mouse wheel values is browser (chromium: 53, firefox: 3)
```
Comment (not-fl3) [inline examples/camera.rs:51]: Maybe do normalization on macroquad side, not on the user code side? 
Not necessarily as a part of this PR though

**Comment by not-fl3** [comment]
Hi! 
I like how the new camera example looks! But maybe keep both, the old one for super-simple camera concept showcase and a new one, with some additional logic behind?

**Comment by saucesaft** [comment]
This pull request looks promising, what can i do to help?

File: examples/camera.rs (hunk 6 above)
Comment (zotho) [reply examples/camera.rs]: Formatted with rustfmt https://github.com/not-fl3/macroquad/pull/25/commits/be5cc78efa12709cc7212939866225aacff9a56b#diff-6a6a8253e0291eb499839f6f1f4c15c3R53

File: examples/camera.rs (hunk 8 above)
Comment (zotho) [reply examples/camera.rs]: Maybe move normalization in [mouse_wheel](https://github.com/not-fl3/macroquad/blob/master/src/lib.rs#L345) function?

**Comment by zotho** [comment]
> Hi!
> I like how the new camera example looks! But maybe keep both, the old one for super-simple camera concept showcase and a new one, with some additional logic behind?

@not-fl3  I moved new code to `camera_transformations` example.

File: examples/camera.rs (hunk 5 above)
Comment (zotho) [reply examples/camera.rs]: It was just to write more compact code. Which is easy to add new cases. See diff: https://github.com/not-fl3/macroquad/pull/25/commits/1bd4e3e6364e4aefc8d29a99b0bf9a0ca128bb03#diff-6a6a8253e0291eb499839f6f1f4c15c3L35
However, I removed it.

File: examples/camera.rs (hunk 7 above)
Comment (zotho) [reply examples/camera.rs]: Label removed
